src/main/java/org/bbop/apollo/web/user/oauth/OauthUserAuthentication.java

```diff
@@ -0,0 +1,253 @@
+			e.printStackTrace();
+		} 
+		catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+	}
+
+	@Override
+	public void generateUserLoginPage(HttpServlet servlet, HttpServletRequest request,
+			HttpServletResponse response) throws ServletException {
+		InputStream in = servlet.getServletContext().getResourceAsStream("/user_interfaces/opauth/login.html");
+		BufferedReader reader = new BufferedReader(new InputStreamReader(in));
+		String line;
+		try {
+			while ((line = reader.readLine()) != null) {
+				response.getOutputStream().println(line);
+			}
+			in.close();
+		} catch (IOException e) {
+			throw new ServletException(e);
+		}
+	}
+
+	@Override
+	public String validateUser(HttpServletRequest request,
+			HttpServletResponse response) throws UserAuthenticationException {
+		
+		// The WebApollo user name
+		String username = null;
+		try {
+			// ---------------
+			// Step1: first, WebApollo redirects the user to the provider 
+			// authentication site to log in and approve access for WebApollo. If 
+			// we are here, it is because the provider has responded with an 
+			// authorization code.  The link that directs the user to the remote 
+			// provider is constructed in the WebContent/user_interfaces/oauth/login.jsp file
+			
+			// get the authorization code
+			String auth_code = request.getParameter("code");
+			String error = request.getParameter("error");
+			String state = request.getParameter("state");
+			
+			// we have an error and should not continue.  return a message for the
+			// user.  If the provider provides an "error_description" then include
+			// that in the error message
+			if (error != null) {
+				String error_description = request.getParameter("error_description");
+				throw new UserAuthenticationException("OAuth error: " + error + ". " + error_description + ". ");
+			}
+
+			// ---------------
+			// Step 2: now that we have an authorization code, we must use it to 
+			// request the access token. There are several types of authentication
+			// available by the Oauth2 protocal:  Authentication Code Flow, 
+			// Implicit Flow, Resource Owner Password Flow, and Client Credentials.
+			// We will use Authentication code Flow. Therefore, we begin by
+			// creating an AuthroizationCodeFlow object
+			JsonFactory json_factory = new JacksonFactory();
+			AuthorizationCodeFlow.Builder acfb= new AuthorizationCodeFlow.Builder(
+				BearerToken.authorizationHeaderAccessMethod(),
+				new NetHttpTransport(),
+				json_factory,
+				new GenericUrl(Token_URL),
+				new ClientParametersAuthentication(Client_ID, Client_Secret), 
+				Client_ID, 
+				Token_URL
+			);
+			AuthorizationCodeFlow codeFlow = acfb.setScopes(Arrays.asList(Uname_Field)).build();
+			
+			// Next we need to construct a token request object to request the access token
+			String redirect_uri = "http://localhost:8080/WebApollo/Login?operation=login&forceRedirect=true";
```

Comment:
need to generalize to the server context (not always WebApollo on 8080), but otherwise correct